File: common/src/main/java/com/ultreon/mods/advanceddebug/util/TargetUtils.java
package com.ultreon.mods.advanceddebug.util;

import net.minecraft.client.Minecraft;
import net.minecraft.client.player.LocalPlayer;
import net.minecraft.util.Mth;
import net.minecraft.world.entity.projectile.ProjectileUtil;
import net.minecraft.world.level.ClipContext;
import net.minecraft.world.level.material.FluidState;
import net.minecraft.world.phys.BlockHitResult;
import net.minecraft.world.phys.EntityHitResult;
import net.minecraft.world.phys.HitResult;
import net.minecraft.world.phys.Vec3;
import org.jetbrains.annotations.Nullable;

public class TargetUtils {
    private TargetUtils() {
        throw new UnsupportedOperationException("Cannot instantiate a utility class");
    }

    @Nullable
    public static BlockHitResult block() {
        if (Minecraft.getInstance().player != null) {
            LocalPlayer player = Minecraft.getInstance().player;
            float xRot = player.getXRot();
            float yRot = player.getYRot();

            Vec3 startPos = player.getEyePosition(1f);

            float calcZ = Mth.cos(-yRot * ((float) Math.PI / 180f) - (float) Math.PI);
            float calcX = Mth.sin(-yRot * ((float) Math.PI / 180f) - (float) Math.PI);
            float rot = -Mth.cos(-xRot * ((float) Math.PI / 180f));
            float deltaY = Mth.sin(-xRot * ((float) Math.PI / 180f));

            float deltaX = calcX * rot;
            float deltaZ = calcZ * rot;

            double distance = 16;

            Vec3 endPos = startPos.add((double) deltaX * distance, (double) deltaY * distance, (double) deltaZ * distance);

            if (Minecraft.getInstance().level != null) {
                return Minecraft.getInstance().level.clip(new ClipContext(startPos, endPos, ClipContext.Block.OUTLINE, ClipContext.Fluid.NONE, player));
            }
        }
        return null;
    }

    @Nullable
    public static BlockHitResult fluid() {
        Minecraft mc = Minecraft.getInstance();
        if (mc.player != null) {
            LocalPlayer player = mc.player;
            float xRot = player.getXRot();
            float yRot = player.getYRot();

            Vec3 startPos = player.getEyePosition(1f);

            float calcZ = Mth.cos(-yRot * ((float) Math.PI / 180f) - (float) Math.PI);
            float calcX = Mth.sin(-yRot * ((float) Math.PI / 180f) - (float) Math.PI);
            float rot = -Mth.cos(-xRot * ((float) Math.PI / 180f));
            float deltaY = Mth.sin(-xRot * ((float) Math.PI / 180f));

            float deltaX = calcX * rot;
            float deltaZ = calcZ * rot;

            double distance = 16;

            Vec3 endPos = startPos.add((double) deltaX * distance, (double) deltaY * distance, (double) deltaZ * distance);

            if (mc.level != null) {
                BlockHitResult clip = mc.level.clip(new ClipContext(startPos, endPos, ClipContext.Block.OUTLINE, ClipContext.Fluid.ANY, player));
                FluidState fluidState = mc.level.getFluidState(clip.getBlockPos());
                if (fluidState.isEmpty()) {
                    return null;
                }
                return clip;
            }
        }
        return null;
    }

    @Nullable
    public static EntityHitResult entity() {
        if (Minecraft.getInstance().player != null) {
            LocalPlayer player = Minecraft.getInstance().player;
            float xRot = player.getXRot();
            float yRot = player.getYRot();

            Vec3 startPos = player.getEyePosition(1f);

            float calcZ = Mth.cos(-yRot * ((float) Math.PI / 180f) - (float) Math.PI);
            float calcX = Mth.sin(-yRot * ((float) Math.PI / 180f) - (float) Math.PI);
            float rot = -Mth.cos(-xRot * ((float) Math.PI / 180f));
            float deltaY = Mth.sin(-xRot * ((float) Math.PI / 180f));

            float deltaX = calcX * rot;
            float deltaZ = calcZ * rot;

            double distance = 16;

            Vec3 endPos = startPos.add((double) deltaX * distance, (double) deltaY * distance, (double) deltaZ * distance);

            if (Minecraft.getInstance().level != null) {
                HitResult hit = Minecraft.getInstance().level.clip(new ClipContext(startPos, endPos, ClipContext.Block.COLLIDER, ClipContext.Fluid.NONE, player));
                if (hit.getType() != HitResult.Type.MISS) {
                    endPos = hit.getLocation();
                }

                HitResult tempHit = ProjectileUtil.getEntityHitResult(Minecraft.getInstance().level, player, startPos, endPos, player.getBoundingBox().inflate(16.0D), entity -> !entity.equals(player));
                if (tempHit != null) {
                    hit = tempHit;
                }
                if (hit instanceof EntityHitResult entityHit) {
                    return entityHit;
                }
            }
        }
        return null;
    }
}

File: common/src/main/kotlin/com/ultreon/mods/advanceddebug/util/TargetUtils.kt
package com.ultreon.mods.advanceddebug.util

import net.minecraft.client.Minecraft
import net.minecraft.util.Mth
import net.minecraft.world.entity.Entity
import net.minecraft.world.entity.projectile.ProjectileUtil
import net.minecraft.world.level.ClipContext
import net.minecraft.world.phys.BlockHitResult
import net.minecraft.world.phys.EntityHitResult
import net.minecraft.world.phys.HitResult

class TargetUtils private constructor() {
    init {
        throw UnsupportedOperationException("Cannot instantiate a utility class")
    }

    companion object {
        fun block(): BlockHitResult? {
            if (Minecraft.getInstance().player != null) {
                val player = Minecraft.getInstance().player
                val xRot = player!!.xRot
                val yRot = player.yRot
                val startPos = player.getEyePosition(1f)
                val calcZ = Mth.cos(-yRot * (Math.PI.toFloat() / 180f) - Math.PI.toFloat())
                val calcX = Mth.sin(-yRot * (Math.PI.toFloat() / 180f) - Math.PI.toFloat())
                val rot = -Mth.cos(-xRot * (Math.PI.toFloat() / 180f))
                val deltaY = Mth.sin(-xRot * (Math.PI.toFloat() / 180f))
                val deltaX = calcX * rot
                val deltaZ = calcZ * rot
                val distance = 16.0
                val endPos = startPos.add(
                    deltaX.toDouble() * distance,
                    deltaY.toDouble() * distance,
                    deltaZ.toDouble() * distance
                )
                if (Minecraft.getInstance().level != null) {
                    return Minecraft.getInstance().level!!.clip(
                        ClipContext(
                            startPos,
                            endPos,
                            ClipContext.Block.OUTLINE,
                            ClipContext.Fluid.NONE,
                            player
                        )
                    )
                }
            }
            return null
        }

        fun fluid(): BlockHitResult? {
            val mc = Minecraft.getInstance()
            if (mc.player != null) {
                val player = mc.player
                val xRot = player!!.xRot
                val yRot = player.yRot
                val startPos = player.getEyePosition(1f)
                val calcZ = Mth.cos(-yRot * (Math.PI.toFloat() / 180f) - Math.PI.toFloat())
                val calcX = Mth.sin(-yRot * (Math.PI.toFloat() / 180f) - Math.PI.toFloat())
                val rot = -Mth.cos(-xRot * (Math.PI.toFloat() / 180f))
                val deltaY = Mth.sin(-xRot * (Math.PI.toFloat() / 180f))
                val deltaX = calcX * rot
                val deltaZ = calcZ * rot
                val distance = 16.0
                val endPos = startPos.add(
                    deltaX.toDouble() * distance,
                    deltaY.toDouble() * distance,
                    deltaZ.toDouble() * distance
                )
                if (mc.level != null) {
                    val clip = mc.level!!.clip(
                        ClipContext(
                            startPos,
                            endPos,
                            ClipContext.Block.OUTLINE,
                            ClipContext.Fluid.ANY,
                            player
                        )
                    )
                    val fluidState = mc.level!!.getFluidState(clip.blockPos)
                    return if (fluidState.isEmpty) {
                        null
                    } else clip
                }
            }
            return null
        }

        fun entity(): EntityHitResult? {
            if (Minecraft.getInstance().player != null) {
                val player = Minecraft.getInstance().player
                val xRot = player!!.xRot
                val yRot = player.yRot
                val startPos = player.getEyePosition(1f)
                val calcZ = Mth.cos(-yRot * (Math.PI.toFloat() / 180f) - Math.PI.toFloat())
                val calcX = Mth.sin(-yRot * (Math.PI.toFloat() / 180f) - Math.PI.toFloat())
                val rot = -Mth.cos(-xRot * (Math.PI.toFloat() / 180f))
                val deltaY = Mth.sin(-xRot * (Math.PI.toFloat() / 180f))
                val deltaX = calcX * rot
                val deltaZ = calcZ * rot
                val distance = 16.0
                var endPos = startPos.add(
                    deltaX.toDouble() * distance,
                    deltaY.toDouble() * distance,
                    deltaZ.toDouble() * distance
                )
                if (Minecraft.getInstance().level != null) {
                    var hit: HitResult = Minecraft.getInstance().level!!.clip(
                        ClipContext(
                            startPos,
                            endPos,
                            ClipContext.Block.COLLIDER,
                            ClipContext.Fluid.NONE,
                            player
                        )
                    )
                    if (hit.type != HitResult.Type.MISS) {
                        endPos = hit.location
                    }
                    val tempHit: HitResult? = ProjectileUtil.getEntityHitResult(
                        Minecraft.getInstance().level!!,
                        player,
                        startPos,
                        endPos,
                        player.boundingBox.inflate(16.0)
                    ) { entity: Entity -> entity != player }
                    if (tempHit != null) {
                        hit = tempHit
                    }
                    if (hit is EntityHitResult) {
                        return hit
                    }
                }
            }
            return null
        }
    }
}